### PR TITLE
[AArch64] Generate zeroing forms of certain SVE2.2 instructions (1/11)

### PR DESCRIPTION
**Pull request by momchil-velikov**
SVE2.2 introduces instructions with predicated forms with zeroing of the inactive lanes. This allows in some cases to save a `movprfx` or a `mov` instruction when emitting code for `_x` or `_z` variants of intrinsics.

This patch adds support for emitting the zeroing forms of `ABS`, `NEG`, `FABS`, and `FNEG` instructions.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 1]
```diff
@@ -381,6 +381,9 @@ def NoUseScalarIncVL : Predicate<"!Subtarget->useScalarIncVL()">;
 
 def UseSVEFPLD1R : Predicate<"!Subtarget->noSVEFPLD1R()">;
 
+def UseUnaryUndefPseudos
+    : Predicate<"!(Subtarget->hasSVE2() || (Subtarget->isStreaming() && Subtarget->hasSME2()))">;
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:385]: This looks like  `!HasSVE2orSME2` rather than the intended `!HasSVE2p2orSME2p2`?

File: llvm/lib/Target/AArch64/SVEInstrFormats.td [hunk 2]
```diff
@@ -517,13 +524,12 @@ multiclass SVE_1_Op_PassthruUndef_Round_Pat<ValueType vtd, SDPatternOperator op,
...
 }
 
-def SVEDup0 : ComplexPattern<vAny, 0, "SelectDupZero", []>;
-
 class SVE_1_Op_PassthruZero_Pat<ValueType vtd, SDPatternOperator op, ValueType vt1,
                    ValueType vt2, Instruction inst>
    : Pat<(vtd (op (vtd (SVEDup0)), vt1:$Op1, vt2:$Op2)),
         (inst (IMPLICIT_DEF), $Op1, $Op2)>;
 
+
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/SVEInstrFormats.td:532]: unnecessary change.

File: llvm/lib/Target/AArch64/SVEInstrFormats.td [hunk 3]
```diff
@@ -4950,7 +4949,22 @@ multiclass sve_int_un_pred_arit_bitwise_fp<bits<3> opc, string asm,
...
 
+  def : SVE_1_Op_PassthruUndefZero_Pat<nxv8f16, op, nxv8i1, nxv8f16, !cast<Instruction>(NAME # _H)>;
+  def : SVE_1_Op_PassthruUndefZero_Pat<nxv4f16, op, nxv4i1, nxv4f16, !cast<Instruction>(NAME # _H)>;
+  def : SVE_1_Op_PassthruUndefZero_Pat<nxv2f16, op, nxv2i1, nxv2f16, !cast<Instruction>(NAME # _H)>;
+  def : SVE_1_Op_PassthruUndefZero_Pat<nxv4f32, op, nxv4i1, nxv4f32, !cast<Instruction>(NAME # _S)>;
+  def : SVE_1_Op_PassthruUndefZero_Pat<nxv2f32, op, nxv2i1, nxv2f32, !cast<Instruction>(NAME # _S)>;
+  def : SVE_1_Op_PassthruUndefZero_Pat<nxv2f64, op, nxv2i1, nxv2f64, !cast<Instruction>(NAME # _D)>;
+}
+
+multiclass sve_int_un_pred_arit_hsd<SDPatternOperator op> {
```
Comment (paulwalker-arm) [inline llvm/lib/Target/AArch64/SVEInstrFormats.td:4967]: I suggest `sve_fp_un_pred_arit_hsd`.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td (hunk 1 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td]: Fixed.

File: llvm/lib/Target/AArch64/SVEInstrFormats.td (hunk 2 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/SVEInstrFormats.td]: Fixed.

File: llvm/lib/Target/AArch64/SVEInstrFormats.td (hunk 3 above)
Comment (momchil-velikov) [reply llvm/lib/Target/AArch64/SVEInstrFormats.td]: Done.

**Review by paulwalker-arm — APPROVED** [review]

**Review by jthackray — APPROVED** [review]
LGTM
